File: apps/cli/src/desktop.ts
import { existsSync } from 'node:fs';
import { homedir, platform } from 'node:os';
import { join } from 'node:path';
import { spawn } from 'node:child_process';

const APP_NAME = 'otto';

function getDesktopAppPath(): string | null {
	const os = platform();

	if (os === 'darwin') {
		const candidates = [
			`/Applications/${APP_NAME}.app`,
			join(homedir(), 'Applications', `${APP_NAME}.app`),
		];
		for (const p of candidates) {
			if (existsSync(p)) return p;
		}
	}

	if (os === 'linux') {
		const candidates = [
			'/usr/bin/otto-desktop',
			'/usr/local/bin/otto-desktop',
			join(homedir(), '.local', 'bin', 'otto-desktop'),
			'/opt/otto-desktop/otto-desktop',
		];
		for (const p of candidates) {
			if (existsSync(p)) return p;
		}
	}

	if (os === 'win32') {
		const candidates = [
			join(
				process.env.LOCALAPPDATA || '',
				'Programs',
				APP_NAME,
				`${APP_NAME}.exe`,
			),
			join(
				process.env.PROGRAMFILES || 'C:\\Program Files',
				APP_NAME,
				`${APP_NAME}.exe`,
			),
		];
		for (const p of candidates) {
			if (existsSync(p)) return p;
		}
	}

	return null;
}

export function isDesktopInstalled(): boolean {
	return getDesktopAppPath() !== null;
}

export function openDesktop(projectPath: string): boolean {
	const appPath = getDesktopAppPath();
	if (!appPath) return false;

	const os = platform();

	try {
		if (os === 'darwin') {
			spawn('open', [appPath, '--args', '--project', projectPath], {
				detached: true,
				stdio: 'ignore',
			}).unref();
		} else if (os === 'win32') {
			spawn(appPath, ['--project', projectPath], {
				detached: true,
				stdio: 'ignore',
			}).unref();
		} else {
			spawn(appPath, ['--project', projectPath], {
				detached: true,
				stdio: 'ignore',
			}).unref();
		}
		return true;
	} catch {
		return false;
	}
}

File: apps/launcher/src-tauri/src/commands/desktop.rs
use std::path::PathBuf;
use std::process::Command;

fn get_desktop_app_path() -> Option<PathBuf> {
    let os = std::env::consts::OS;

    if os == "macos" {
        let candidates = vec![
            PathBuf::from("/Applications/otto.app"),
            dirs::home_dir()
                .map(|h| h.join("Applications").join("otto.app"))
                .unwrap_or_default(),
        ];
        for p in candidates {
            if p.exists() {
                return Some(p);
            }
        }
    }

    if os == "linux" {
        let candidates = vec![
            PathBuf::from("/usr/bin/otto-desktop"),
            PathBuf::from("/usr/local/bin/otto-desktop"),
            dirs::home_dir()
                .map(|h| h.join(".local").join("bin").join("otto-desktop"))
                .unwrap_or_default(),
            PathBuf::from("/opt/otto-desktop/otto-desktop"),
        ];
        for p in candidates {
            if p.exists() {
                return Some(p);
            }
        }
    }

    None
}

#[tauri::command]
pub fn is_desktop_installed() -> bool {
    get_desktop_app_path().is_some()
}

#[tauri::command]
pub fn open_in_desktop(api_url: String, name: String) -> Result<(), String> {
    let app_path = get_desktop_app_path().ok_or("Desktop app not found")?;
    let os = std::env::consts::OS;

    if os == "macos" {
        Command::new("open")
            .args([
                "-n",
                app_path.to_str().unwrap_or_default(),
                "--args",
                "--remote",
                &api_url,
                "--name",
                &name,
            ])
            .spawn()
            .map_err(|e| e.to_string())?;
    } else {
        Command::new(app_path)
            .args(["--remote", &api_url, "--name", &name])
            .spawn()
            .map_err(|e| e.to_string())?;
    }

    Ok(())
}
